netket/logging.py

```diff
@@ -0,0 +1,115 @@
+import json as _json
+import netket as _nk
+from os import path as _path
+
+
+def _exists_json(prefix):
+    return _path.exists(prefix + ".log") or _path.exists(prefix + ".wf")
+
+
+class JsonLog:
+    """
+    Creates a Json Logger sink object, that can be passed with keyword argument `logger` to Monte
+    Carlo drivers in order to serialize the outpit data of the simulation.
+
+    Args:
+        output_prefix: the name of the output files before the extension
+        save_params_every: every how many iterations should machine parameters be flushed to file
+        write_every: every how many iterations should data be flushed to file
+        mode: Specify the behaviour in case the file already exists at this output_prefix. Options
+        are `[o]verwrite` (default), `[a]ppend`, `[n]ewfile` (creates a new file with an incremental
+        name) or `[f]ail`.
+    """
+
+    def __init__(
+        self, output_prefix, save_params_every=50, write_every=50, mode="overwrite"
+    ):
+        # Shorthands for mode
+        if mode == "o":
+            mode = "overwrite"
+        elif mode == "a":
+            mode = "append"
+        elif mode == "n":
+            mode = "newfile"
+        elif mode == "f":
+            mode = "fail"
+
+        if not (
+            (mode == "overwrite")
+            or (mode == "append")
+            or (mode == "newfile")
+            or (mode == "fail")
+        ):
+            raise ValueError(
+                "Mode not recognized: should be one of `[o]verwrite`, `[a]ppend`, `[n]ewfile or `[f]ail`."
+            )
+
+        file_exists = _exists_json(output_prefix)
+
+        starting_json_content = {"Output": []}
+
+        if file_exists and mode == "newfile":
+            i = 1
+            # Generate a new output_prefix with an incremental counter
+            while i < 1001:
+                _prefix = output_prefix + "_{}".format(i)
+                if not _exists_json(_prefix):
+                    output_prefix = _prefix
+                    break
+                i = i + 1
+
+            # give up after we tried too many files to avoid an infinite loop
+            if i is 1000:
+                raise IOError(
+                    "More than 1000 files exist with that prefix. Please change ouput_prefix manually."
+                )
```

Comment:
Not sure whether this mode needs to be part of the logger code. I believe it's the responsibility of user code to provide the file name that should actually be used. We could provide this code here as a helper function, so that user who want his could call (my name might be too long - but descriptive :slightly_smiling_face:)
```python
prefix = netket.logging.increment_prefix_if_needed("myprefix")
vmc.run(..., prefix)
```

Reply:
So basically this is a choice between your syntax
```python
prefix = netket.logging.increment_prefix_if_needed("myprefix")
logger = JsonLogger(prefix)
```
and my syntax
```python
logger = JsonLogger(prefix, mode='newfile')
```
I am personally usually in favour of terse syntax. And in the future this can be moved to an `AbstractLogger` class.

It's not just a stylistic convenience: this same syntax is used by the TensorBoard loggers for Python (TensorBoardX) and Julia (TensorBoardLogger). By the way, In the latter it's the default behaviour, and I know people appreciate the fact that it's non-destructive by default.

Reply:
Not sure something like `mode` should be part of an abstract logger interface, the meaning of it strongly depends on the output type (e.g., there could be a logger that writes to a remote database which has quite different semantics from a local file).

One issue I have with the current way: If I run a script that uses a constant prefix, say `test`, every time, I get the output files `test.log`, `test_1.log`, `test_2.log`, right? The first file would always break the pattern, which in my experience can be a bit annoying. User code however should know this and would be a good place to ensure a consistent naming for output.

Reply:
ok.
I'll remove this code for now, and leave the other three modes.